src/goto-programs/goto_clean_expr.cpp

```diff
@@ -405,23 +387,21 @@ void goto_convertt::clean_expr(
   }
   else if(expr.id()==ID_forall || expr.id()==ID_exists)
   {
-    assert(expr.operands().size()==2);
+    quantifier_exprt &quatifier_expr = to_quantifier_expr(expr);
     // check if there are side-effects
     goto_programt tmp;
-    clean_expr(expr.op1(), tmp, mode, true);
-    if(tmp.instructions.empty())
-    {
-      error().source_location=expr.find_source_location();
-      error() << "no side-effects in quantified expressions allowed"
-              << eom;
-      throw 0;
-    }
+    clean_expr(quatifier_expr.where(), tmp, mode, true);
+
+    INVARIANT_WITH_DIAGNOSTICS(
+      !tmp.instructions.empty(),
+      "quantifier expressions cannot have side-effects",
+      expr.find_source_location());
     return;
```

Comment:
This is a case where user-input is checked; i.e., this should not be an invariant.

Reply:
Since a3fa91130b2b0d93cc1, the front-end checks whether quantifiers contain side-effects; i.e., this is now an invariant!